File: 2026/02-February/03-02-26/0050-powx-n.cpp
/*
 * LeetCode 50 · Pow(x, n) · Medium
 * https://leetcode.com/problems/powx-n/
 *
 * Pattern : Math & Logic
 * Solved  : 03 Feb 2026
 * Time    : O(n)
 * Space   : O(1)
 */

class Solution {
public:
    double pow(double x,long long n,double w){
        if(n==0) return w;
        return pow(x,n-1,w*x);
    }
    double p(double x,long long n,double w){
        if(n==0) return w;
        return (p(x,n+1,w*x));
    }
    double myPow(double x, long long  n) {
        if(n==0) return 1;
        if(x==1) return 1;
        if(x== -1 ){

            if(n%2!=0){
                return -1;
            }
            else return 1;
        }
        if(n>0){
            double r=pow(x,n/2,1);
            if(n % 2 == 0)
                return r * r;
            else
                return r * r * x;   
        }
        else{
            double rr= p(x,n/2,1);
            if(n % 2 == 0)
                return 1/(rr * rr);
            else
                return 1/(rr * rr * x);   

        }
        
        
        
    }
};

File: 2026/09-September/23-09-26/0050-powx-n.py
"""
LeetCode 50 · Pow(x, n) · Medium
https://leetcode.com/problems/powx-n/

Pattern : Data Structures & Algorithms
Solved  : 23 Sep 2026
Time    : O(n) · Runtime: 0 ms (Beats 100.0%)
Space   : O(1) · Memory: 12.4 MB (Beats 63.2%)
"""

class Solution(object):
    def myPow(self, x, n):
        """
        :type x: float
        :type n: int
        :rtype: float
        """
        if n==0:
            return 1
        if n<0:
            return 1/self.myPow(x,-n)
        
        half= self.myPow(x,n//2)
        if n%2==0:
            return half*half 
        else:
            return half*half*x
